lib/postgrex/notifications.ex

```diff
@@ -74,22 +74,87 @@ defmodule Postgrex.Notifications do
   @doc false
   defstruct idle_interval: 5000,
             protocol: nil,
-            listeners: %{},
-            listener_channels: %{},
             auto_reconnect: false,
             reconnect_backoff: 500,
-            connected: false
+            state: nil
 
   ## PUBLIC API ##
 
+  @type query :: iodata
   @type server :: GenServer.server()
+  @type state :: term
+
+  @doc """
+  Callback for process initialization.
+
+  This is called once and before the Postgrex notification connection is established.
+  """
+  @callback init(term) :: {:ok, state}
+
+  @doc """
+  Invoked after connecting.
+
+  This may be called multiple times if `:auto_reconnect` is true.
+  """
+  @callback handle_connect(state) :: {:noreply, state} | {:query, query, state}
+
+  @doc """
+  Invoked after disconnection, regardless of `:auto_reconnect`.
+  """
+  @callback handle_disconnect(state) :: {:noreply, state}
+
+  @doc """
+  Callback for `call/3`.
+
+  Replies must be sent with `reply/2`.
+  """
+  @callback handle_call(term, GenServer.from(), state) ::
+              {:noreply, state}
+              | {:query, query, state}
+
+  @doc """
+  Callback for `Kernel.send/2`.
+  """
+  @callback handle_info(term, state) :: {:noreply, state} | {:query, query, state}
+
+  @doc """
+  Callback for processing or relaying query results.
+  """
+  @callback handle_result(Postgrex.Result.t() | Postgrex.Error.t(), state) :: {:noreply, state}
+
+  @doc """
+  Callback for processing or relaying pubsub notifications.
+  """
+  @callback handle_notification(binary, binary, state) :: {:noreply, state}
```

Comment:
I think the state in handle_notification is ignored and not given back, right? Perhaps the return type here should be `:ok`. Potentially even call the callback `notify` so people don't expect it to have the same return types as `handle_`.

Reply:
Good call. Changing it to `notify` and `:ok` 👍